File: medium/cpp/366.cpp
// https://leetcode.com/problems/find-leaves-of-binary-tree/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    std::vector<std::vector<int>> findLeaves(TreeNode* root)
    {
        std::vector<std::vector<int>> res;
        static_cast<void>(go(root, res));
        return res;
    }

private:
    int go(TreeNode* root, std::vector<std::vector<int>>& res)
    {
        if (root == nullptr)
        {
            return -1;
        }

        const auto d = std::max(go(root->left, res), go(root->right, res)) + 1;

        if (std::size(res) == d)
        {
            res.push_back({});
        }

        res[d].push_back(root->val);

        return d;
    }
};

File: medium/python/366.py
# https://leetcode.com/problems/find-leaves-of-binary-tree/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findLeaves(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []

        def dfs(root):
            if not root:
                return -1

            d = 1 + max(dfs(root.left), dfs(root.right))

            if len(res) == d:
                res.append([])

            res[d].append(root.val)

            return d

        dfs(root)

        return res
        